lib/Sema/NameBinding.cpp

```diff
@@ -36,60 +37,539 @@
+  };
+
+  class NameBinder : public DeclVisitor<NameBinder> {
     SourceFile &SF;
-    ASTContext &Context;
+    ASTContext &ctx;
 
-    NameBinder(SourceFile &SF) : SF(SF), Context(SF.getASTContext()) {}
+    /// Imports which still need their options checked, modules loaded, and
+    /// cross-imports found.
+    SmallVector<UnboundImport, 4> unboundImports;
 
+    /// Imports which still need their scoped imports validated.
+    SmallVector<BoundImport, 16> unvalidatedImports;
+
+  public:
+    NameBinder(SourceFile &SF)
+      : SF(SF), ctx(SF.getASTContext())
+    { }
+
+    // Special behavior for these decls:
+    void visitImportDecl(ImportDecl *ID);
+    void visitPrecedenceGroupDecl(PrecedenceGroupDecl *group);
+    void visitPrefixOperatorDecl(PrefixOperatorDecl *OpDecl);
+    void visitInfixOperatorDecl(InfixOperatorDecl *OpDecl);
+    void visitPostfixOperatorDecl(PostfixOperatorDecl *OpDecl);
+
+    // Ignore other decls.
+    void visitDecl(Decl *D) {}
+
+    /// Postprocess the imports this NameBinder has bound and collect them into
+    /// a vector.
+    void finishImports(SmallVectorImpl<ImportedModuleDesc> &imports);
+
+  protected:
     template<typename ...ArgTypes>
     InFlightDiagnostic diagnose(ArgTypes &&...Args) {
-      return Context.Diags.diagnose(std::forward<ArgTypes>(Args)...);
+      return ctx.Diags.diagnose(std::forward<ArgTypes>(Args)...);
     }
 
-    void addImport(SmallVectorImpl<SourceFile::ImportedModuleDesc> &imports,
-                   ImportDecl *ID);
+    /// Check a single unbound import, bind it, add it to \c unvalidatedImports,
+    /// and add its cross-import overlays to \c unboundImports.
+    void bindImport(UnboundImport &&I);
+
+    /// Adds \p I and \p M to \c unvalidatedImports and \c visibleModules.
+    void addImport(const UnboundImport &I, ModuleDecl *M,
+                   bool needsScopeValidation);
+
 
     /// Load a module referenced by an import statement.
     ///
     /// Returns null if no module can be loaded.
     ModuleDecl *getModule(ArrayRef<Located<Identifier>> ModuleID);
+
+    template<typename OP_DECL>
+    void insertOperatorDecl(SourceFile::OperatorMap<OP_DECL*> &Operators,
+                            OP_DECL *OpDecl);
   };
 } // end anonymous namespace
 
-ModuleDecl *
-NameBinder::getModule(ArrayRef<Located<Identifier>> modulePath) {
+//===----------------------------------------------------------------------===//
+// MARK: performNameBinding
+//===----------------------------------------------------------------------===//
+
+/// performNameBinding - Once parsing is complete, this walks the AST to
+/// resolve names and do other top-level validation.
+///
+/// At this point parsing has been performed, but we still have
+/// UnresolvedDeclRefExpr nodes for unresolved value names, and we may have
+/// unresolved type names as well. This handles import directives and forward
+/// references.
+void swift::performNameBinding(SourceFile &SF) {
```

Comment:
I don't see a connection between the first sentence and the second sentence in the doc comment's paragraph

```
 /// At this point parsing has been performed, but we still have
 /// UnresolvedDeclRefExpr nodes for unresolved value names, and we may have
 /// unresolved type names as well. This handles import directives and forward
 /// references.
```

Are they related? If not, can you separate the second sentence into a separate paragraph? Something like:

```
 /// At the time this function is called, parsing has been performed, but we still have
 /// UnresolvedDeclRefExpr nodes for unresolved value names, and we may have
 /// unresolved type names as well. Those names are resolved.
 ///
 /// We also handle import directives and forward references here.
```

Reply:
Oh looks like this was not part of your PR. Nvm, we can fix this later.

Reply:
I think the first sentence is a dead comment. My impression is that, once upon a time, NameBinding.cpp bound *all* of the names; now it just binds a few that don't work through normal lookup.

Reply:
```
/// performNameBinding - Once parsing is complete, this walks the AST to
/// resolve names and do other top-level validation.
///
/// Most names are actually bound by the type checker, but before we can
/// type-check a source file, we need to make declarations imported from other
/// modules available and build tables of the operators and precedecence groups
/// declared in that file. Name binding processes top-level \c ImportDecl,
/// \c OperatorDecl, and \c PrecedenceGroupDecl nodes to perform these tasks,
/// along with related validation.
///
/// Name binding operates on a parsed but otherwise unvalidated AST.
```